File: DataStructure/146.lru-cache.go
package leetcode

type n struct {
	key, val   int
	next, prev *n
}
type LRUCache struct {
	capacity int
	m        map[int]*n
	dummy    *n
}

func Constructor146(capacity int) LRUCache {
	dummy := &n{}
	dummy.next, dummy.prev = dummy, dummy
	return LRUCache{capacity, map[int]*n{}, dummy}
}

func (this *LRUCache) Get(key int) int {
	if v, ok := this.m[key]; ok {
		this.detach(v)
		this.leftAppend(v)
		return v.val
	} else {
		return -1
	}
}

func (this *LRUCache) Put(key int, value int) {
	if v, ok := this.m[key]; ok {
		v.val = value
		this.detach(v)
		this.leftAppend(v)
		return
	}
	newN := &n{key: key, val: value}
	this.m[key] = newN
	this.leftAppend(newN)
	if len(this.m) > this.capacity {
		lastItem := this.dummy.prev
		this.detach(lastItem)
		delete(this.m, lastItem.key)
	}
}

func (this *LRUCache) detach(x *n) {
	x.prev.next = x.next
	x.next.prev = x.prev
	x.next, x.prev = nil, nil
}

func (this *LRUCache) leftAppend(x *n) {
	x.prev = this.dummy
	x.next = this.dummy.next
	x.prev.next = x
	x.next.prev = x
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

File: DataStructure/146.lru-cache.py
# @leet start
class node:
    def __init__(self, key, value, next=None, prev=None) -> None:
        self.value = value
        self.key = key
        self.next = next
        self.prev = prev

    def detach(self):
        self.prev.next = self.next  # type: ignore
        self.next.prev = self.prev  # type: ignore
        self.prev = self.next = None


class LRUCache:
    def __init__(self, capacity: int):
        dummy = node(0, 0)
        dummy.next = dummy.prev = dummy
        self.dummy = dummy
        self.capacity = capacity
        self.m = {}

    def appendLeft(self, x):
        x.prev = self.dummy
        x.next = self.dummy.next
        x.prev.next = x
        x.next.prev = x  # type: ignore

    def getNode(self, key):
        if key in self.m:
            x = self.m[key]
            x.detach()
            self.appendLeft(x)
            return x
        return None

    def get(self, key: int) -> int:
        n = self.getNode(key)
        if n:
            return n.value
        return -1

    def put(self, key: int, value: int) -> None:
        n = self.getNode(key)
        if n:
            n.value = value
            return

        if len(self.m) == self.capacity:
            tail = self.dummy.prev
            tail.detach()  # type: ignore
            del self.m[tail.key]  # type: ignore

        nn = node(key, value)
        self.m[key] = nn
        self.appendLeft(nn)

